Offline Mode: Prepublishing sheet fixes

kean opened the pull request:
- Add `DeprecatedPrepublishingViewController`. There were too many changes required in the screen and it was no longer sustainable to keep adding if statement to the exiting screen. The deprecated version now fully matches the production version to ensure we don't regress the experience for users who don't have this feature flag on.
- Fix an issue with posts being published as "protected" by default
- (Revert) Remove "Publish Date" from the draft Post Settings again (it's unfixable without the switch to the .org API). Related discussion about the publish date: p1712869221226789-slack-C06GRKUGDNX. The republishing sheet will now require you to set the "publish date" right before publishing.
- Fix an issue where a standalone prepublishing sheet would not send the tags and categories when publishing (it required a revision)
- Fix an issue where the changes made in the standalone prepublishing sheet would be saved to the DB and not the server
- Fix an issue with publish button label not updating after changing visibility to .private

Related WPKit PR https://github.com/wordpress-mobile/WordPressKit-iOS/pull/786
 
## To test:

**Test 1.1**

- Create and save a draft post
- Open Post Settings
- **Verify** that the "Publish Date" says "Immediately"
- Change the publish date to a different date
- Return back to Post Settings and verify that the date was displayed
- Save the changes
- Reopen Post Settings and **verify** that is still shows the selected date
- Remove the publish date (in the current branch, tap "Now")
- **Verify** that the "Publish Date" says "Immediately"
- Save 
- Reopend Post Settings and **verify** that the date still says now

**Test 2.1**

- Create and save a draft post
- Open Post Settings and save the publish date in the future
- Tap "Publish" from the list
- Select a future date in the past
- Tap "Publish" to confirm
- **Verify** that the post got published with the selected date

**Test 2.2**

- Create and save a draft post
- Tap "Publish" from the List
- Add a tag
- Tap "Publish" again to confirm
- **Verify** that the post was published with the selected tag

**Test 2.3**

- Create and save a draft post
- Tap "Publish" from the List
- Add a tag
- Tap "Close"
- **Verify** that the revision was deleted and the changes were not saved

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

kean committed:
Fix an issue with PostCoordinator always sending a password field

kean committed:
Remove Publish Date option from Post Settings for draft and pending posts

kean committed:
Update comments

kean committed:
Remove schedule action from PostEditorState

kean committed:
Add PrepublishingViewModel

kean committed:
Add DeprecatedPrepublishingViewController

kean committed:
Rework the prepublishing sheet

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean committed:
Fix publish date formatting when .syncPublishing is off

momo-ozawa reviewed (commented):
### Test 1.1
* FF disabled: The post gets published if you change from a future date to "Now". Similar to [this bug](https://github.com/wordpress-mobile/WordPress-iOS/pull/22930#pullrequestreview-1976510905)
* FF enabled: N/A since "Publish Date" has been removed from Post Settings

### Test 2.1
![IMG_0039](https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/e48f9e67-aaf5-4049-bc58-03db8eaa9e89)

* FF disabled: I don't see a "Publish" action in the context menu for scheduled posts. If I set the publish date to the future, the post gets moved to the scheduled tab. 
* FF enabled: N/A since "Publish Date" has been removed from Post Settings

### Test 2.2
Before setting tag | After setting tag
-- | --
![IMG_0041](https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/a0e19f98-3e00-490b-b0c3-a336c2058060) | ![IMG_0040](https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/f8bf0961-9ee2-46c2-8bae-f0046ee29255)

* FF disabled/enabled: The scrollview height is wrong after setting a tag

### Test 2.3

https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/8d0de58c-f6d2-4ac8-95a9-c52afb21273d



* FF disabled: After closing the prepublishing sheet, the first time I open the post settings from the context menu, I still see the tag I added. The second time I open the post settings, the tag has been removed.

kean commented:
> FF disabled: The post gets published if you change from a future date to "Now". Similar to https://github.com/wordpress-mobile/WordPress-iOS/pull/22930#pullrequestreview-1976510905

It's a production issue (tested in 24.6).

> FF disabled: I don't see a "Publish" action in the context menu for scheduled posts. If I set the publish date to the future, the post gets moved to the scheduled tab.

Yeah, I think I wrote this test based on the assumption without confirming it. But it doesn't work this way because of the known issue in test 1.1.

> FF disabled/enabled: The scrollview height is wrong after setting a tag

I added it to the list of known issues. I've seen it occasionally happen before.

> FF disabled: After closing the prepublishing sheet, the first time I open the post settings from the context menu, I still see the tag I added. The second time I open the post settings, the tag has been removed.


Expected – it's a production issue. The first time you close the settings it calls the following line and reset the changes you made:

```
 [self.apost.managedObjectContext refreshObject:self.apost mergeChanges:NO];
```

kean committed:
Merge branch 'feature/offline-mode-milesone-2' into task/offline-mode-fix-publish-date-handling

kean committed:
Fix race condition in EditorGutenbergTests
